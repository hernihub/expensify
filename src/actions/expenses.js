import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD EXPENSE action generator
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => { // this functions gets called internally by Redux with dispatch as an argument
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return database.ref('expenses').push(expense).then((ref) => { // Returning the promise to it can be chained downstream
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    });
  };
};
  
  // REMOVE EXPENSE action generator
export const removeExpense = ( { id } = {} ) => ({
  type: 'REMOVE_EXPENSE',
  id
});
  
// EDIT EXPENSE action generator
export const editExpense = ( id, updates ) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// Fetch the data and dispatch the latter
export const startSetExpenses = () => {
  return (dispatch) => { // this functions gets called internally by Redux with dispatch as an argument
    return database.ref('expenses').once('value').then((snapshot) => { // Returning the promise to it can be chained downstream
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      })
      dispatch(setExpenses(expenses));
    });
  };
};