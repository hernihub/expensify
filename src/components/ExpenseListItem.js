import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({id, description, amount, createdAt}) => ( // connect function injects dispatch function into the object as a prop
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>{amount} - {createdAt}</p>
  </div>
);

export default ExpenseListItem;