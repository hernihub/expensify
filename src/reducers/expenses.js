// in case of complex expenses states, a default object is in order
const expensesReducerDefault = [];

export default (state = expensesReducerDefault, action) => {
    switch (action.type) {
      case 'ADD_EXPENSE':
        return [
          ...state,
          action.expense
        ];
      case 'REMOVE_EXPENSE':
        return state.filter(({ id }) => id !== action.id ); // destructuring the expense
      // return state.filter( (currenExpense) => currenExpense.id !== action.id ); // without destructuring
      case 'EDIT_EXPENSE':
        return state.map((expense) => {
          if(expense.id === action.id) {
            return { // new object with the old expense object spreaded plus some of its properties overriden by putting them after the spreaded object
              ...expense,
              ...action.updates // object spread
            };
          } else {
            return expense;
          };
        });
      case 'SET_EXPENSES':
        return action.expenses;
      default:
        return state;
  }
};