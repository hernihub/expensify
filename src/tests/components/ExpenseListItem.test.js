import React from 'react';
import {shallow} from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('should render ExpenseListItem with first expense in the array', () =>{
  const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
  expect(wrapper).toMatchSnapshot();
});