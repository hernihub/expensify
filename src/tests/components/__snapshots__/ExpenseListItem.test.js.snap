// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with first expense in the array 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      Marijuana
    </h3>
  </Link>
  <p>
    191500
     - 
    0
  </p>
</div>
`;
